lib/xgu/xgu.h

```diff
@@ -0,0 +1,535 @@
+        float r, g, b, a;
+    };
+    float f[4];
+} XguVec4;
+
+
+/* ========================= *
+ *  BEGIN PUSH HELPER BLOCK  *
+ * ========================= */
+
+static inline
+DWORD* push_command(DWORD* p, DWORD command, int nparams) {
+    pb_push(p++, command, nparams);
+    return p;
+}
+
+static inline
+DWORD* push_parameter(DWORD* p, DWORD parameter) {
+    *p++ = parameter;
+    return p;
+}
+
+static inline
+DWORD* push_boolean(DWORD* p, bool enabled) {
+    return push_parameter(p, enabled ? 1 : 0);
+}
+
+static inline
+DWORD* push_command_boolean(DWORD* p, DWORD command, bool enabled) {
+    p = push_command(p, command, 1);
+    p = push_boolean(p, enabled);
+    return p;
+}
+
+static inline
+DWORD* push_float(DWORD* p, float f) {
+    return push_parameter(p, *(uint32_t*)&f);
+}
+
+static inline
+DWORD* push_floats(DWORD* p, float* f, unsigned int count) {
+    for(unsigned int i = 0; i < count; i++) {
+        p = push_float(p, f[i]);
+    }
+    return p;
+}
+
+static inline
+DWORD* push_matrix2x2(DWORD* p, float m[2*2]) {
+    return push_floats(p, m, 2*2);
+}
+
+static inline
+DWORD* push_matrix4x4(DWORD* p, float m[4*4]) {
+    return push_floats(p, m, 4*4);
+}
+
+static inline
+DWORD* push_command_matrix2x2(DWORD* p, DWORD command, float m[2*2]) {
+    p = push_command(p, command, 2*2);
+    p = push_matrix2x2(p, m);
+    return p;
+}
+
+static inline
+DWORD* push_command_matrix4x4(DWORD* p, DWORD command, float m[4*4]) {
+    p = push_command(p, command, 4*4);
+    p = push_matrix4x4(p, m);
+    return p;
+}
+
+static inline
+DWORD* push_command_parameter(DWORD* p, DWORD command, DWORD parameter) {
+    p = push_command(p, command, 1);
+    p = push_parameter(p, parameter);
+    return p;
+}
+
+static inline
+DWORD* push_command_float(DWORD* p, DWORD command, float parameter) {
+    p = push_command(p, command, 1);
+    p = push_float(p, parameter);
+    return p;
+}
+
+
+/* ========================= *
+ * BEGIN XGU FUNCTIONS BLOCK *
+ * ========================= */
+
+inline
+DWORD* xgu_begin(DWORD* p, XguPrimitiveType type) {
+
+    // Force people to use xgu_end instead
+    assert(type != NV097_SET_BEGIN_END_OP_END);
+
+    return push_command_parameter(p, NV097_SET_BEGIN_END, type);
+}
+
+inline
+DWORD* xgu_end(DWORD* p) {
+    return push_command_parameter(p, NV097_SET_BEGIN_END, NV097_SET_BEGIN_END_OP_END);
+}
+
+inline
+DWORD* xgu_set_viewport_offset(DWORD* p, float xw, float yh, float avgDepth) {
+    p = push_command(p, NV097_SET_VIEWPORT_OFFSET, 4);
+    p = push_float(p, xw);
+    p = push_float(p, yh);
+    p = push_float(p, avgDepth);
+    return push_float(p, (avgDepth == 0.0f ? 0.0f : 16777215.0));
+}
+
+inline
+DWORD* xgu_set_viewport_scale(DWORD* p, float w, float h, float znear, float zfar) {
+    p = push_command(p, NV097_SET_VIEWPORT_SCALE, 4);
+    p = push_float(p, w);
+    p = push_float(p, h);
+    p = push_float(p, ((zfar - znear / 2.0) * 16777214));
+    return push_float(p, 65535.0);
```

Comment:
This is a combination of different steps.

Also XGU is a low-level wrapper around the GPU:  Don't do math.

We have 2 options:
- Split this into different functions and remove all internal math.
- Just have it be: `xgu_set_viewport_scale(float x, float y, float z, float w)` (= Don't assign a meaning)

(Same stuff for the `xgu_set_viewport_offset`)

Reply:
`nv_regs.h` doesn't contain values for the different steps so I guess I'll go with option No. 2 for now.